### DISCUSSION TITLE
Unsupported model field type

### DISCUSSION THREAD

**Discussion by lukezsmith** [silviogutierrez/reactivated#292] [Q&A] [answered]
Hi,

Thanks again for the fantastic framework!

I'm having an issue with serialising custom typed model fields via the [djstripe](https://github.com/dj-stripe/dj-stripe) Python library. 
I am getting the following error: `AssertionError: Unsupported model field type <class 'djstripe.fields.StripeIdField'>. This should probably silently return None and allow a custom handler to support the field.`

Below is my template code:

```python
from typing import NamedTuple, List
from reactivated import template, Pick
from djstripe.models import Product

@template
class PricingPage(NamedTuple):
    subscription_list: List[Pick[Product, "id", "name", "caption"]]
    user: Pick[User, "id", "email"]
```

Where the "id" field is the offending field causing the error. Looking in the djstripe source code the field is defined here: https://github.com/dj-stripe/dj-stripe/blob/a0630793b16638936b04039dce6e40ab55266fcf/djstripe/fields.py#L150
```python
class StripeIdField(FieldDeconstructMixin, models.CharField):
    """A field with enough space to hold any stripe ID."""

    def __init__(self, *args, **kwargs):
        """
        Assign default args to this field.

        As per: https://stripe.com/docs/upgrades
        You can safely assume object IDs we generate will never exceed 255
        characters, but you should be able to handle IDs of up to that
        length.
        """
        defaults = {"max_length": 255, "blank": False, "null": False}
        defaults.update(kwargs)
        super().__init__(*args, **defaults)
```

And here is my view fulfilling the template:

```python
from djstripe.models import Product
def pricing_page(request: HttpRequest) -> HttpResponse:
    return templates.PricingPage(subscription_list=Product.objects.all().order_by("id"), user=request.user).render(request)
```
Is there any way to solve this or any workarounds that you would try if you were in this position? I'm not familiar enough with Python's typing system so I am having trouble trying to understand how to proceed with this beyond replicating the djstripe model myself with supported field types. Any help would be greatly appreciated!

**Comment by silviogutierrez** [comment, marked as answer]
@lukezsmith : thanks for the kind words!

Take a look at this PR from another contributor: https://github.com/silviogutierrez/reactivated/pull/291

It shows how to "register" a field type. Give it a go with your own field as early as possible, perhaps where models load. And map to str. Let me know if it works.

**Reply by lukezsmith** [reply]
@silviogutierrez Thanks for this. 
Turns out simply updating the library to the latest alpha release (0.30.2a1685) solved the issue on the id field without any additional changes.

I was however running into a JSONField unsupported type error: `AssertionError: Unsupported type djstripe.Product.metadata`, where this type uses a derivation of JSONField, and is defined in djstripe at https://github.com/dj-stripe/dj-stripe/blob/a0630793b16638936b04039dce6e40ab55266fcf/djstripe/fields.py#L179:
```python
from django.db.models import JSONField as BaseJSONField

class JSONField(FieldDeconstructMixin, BaseJSONField):
    """A field used to define a JSONField value according to djstripe logic."""

    pass
```

Using the specified register fix in `templates.py` with the JSON (which is set to Any in reactivated/serialization/registry.py) mapping solves this issue (albeit with a bit of a pain of parsing the JSON data in React as an unknown type):
```python
from typing import NamedTuple, List
from reactivated import template, Pick
from djstripe.models import Product
from djstripe.fields import JSONField
from reactivated.serialization.registry import register, JSON

register(JSONField)(JSON)

@template
class PricingPage(NamedTuple):
    subscription_list: List[Pick[Product, "id", "name", "caption", "metadata"]]
    user: Pick[User, "id", "email"]
```
 Thanks again for the extremely fast help and fantastic library!

**Reply by silviogutierrez** [reply]
@lukezsmith : great! What do you think it should map to? `unknown` by default?

FYI, I believe you can override what the field maps to like so:

```python
class MyModel(models.Model):
      foo: int = models.CharField(...)
```

So if you provide a better data shape there, you can "type" your JSON if you know what it looks like. Of course, far less flexible then.

**Reply by silviogutierrez** [reply]
Here's the PR that added that: https://github.com/silviogutierrez/reactivated/pull/96
